consensus/src/aggregation/engine.rs

```diff
@@ -759,37 +779,68 @@ impl<
             .for_each(|certificate| {
                 self.confirmed
                     .insert(certificate.item.index, certificate.clone());
             });
-        // Add any acks that haven't resulted in a threshold signature
-        acks = acks
-            .into_iter()
-            .filter(|ack| ack.item.index >= tip && !self.confirmed.contains_key(&ack.item.index))
-            .collect::<Vec<_>>();
-        acks.iter().for_each(|ack| {
-            assert!(self
-                .pending
-                .insert(
-                    ack.item.index,
-                    Pending::Verified(
-                        ack.item.digest,
-                        HashMap::from([(
-                            ack.epoch,
-                            HashMap::from([(ack.signature.index, ack.clone())]),
-                        )]),
-                    ),
-                )
-                .is_none());
-            self.set_rebroadcast_deadline(ack.item.index);
-        });
+
+        // Group acks by index
+        let mut acks_by_index: BTreeMap<Index, Vec<Ack<V, D>>> = BTreeMap::new();
+        for ack in acks {
+            if ack.item.index >= prune_threshold && !self.confirmed.contains_key(&ack.item.index) {
+                acks_by_index.entry(ack.item.index).or_default().push(ack);
+            }
+        }
+
+        // Process each index's acks
+        for (index, mut acks_group) in acks_by_index {
+            // Check if we have our own ack (which means we've verified the digest)
+            let our_share = self.validators.share(self.epoch);
+            let our_digest = if let Some(share) = our_share {
+                acks_group
+                    .iter()
+                    .find(|ack| ack.epoch == self.epoch && ack.signature.index == share.index)
+                    .map(|ack| ack.item.digest)
+            } else {
+                None
+            };
+
+            // If our_digest exists, delete everything from acks_group that doesn't match it
+            if let Some(digest) = our_digest {
+                acks_group.retain(|ack| ack.item.digest == digest);
+            }
+
+            // Create a new epoch map
+            let mut epoch_map = BTreeMap::new();
+            for ack in acks_group {
+                epoch_map
+                    .entry(ack.epoch)
+                    .or_insert_with(BTreeMap::new)
+                    .insert(ack.signature.index, ack);
+            }
+
+            // Insert as Verified if we have our own ack (meaning we verified the digest),
+            // otherwise as Unverified
+            match our_digest {
+                Some(digest) => {
+                    self.pending
+                        .insert(index, Pending::Verified(digest, epoch_map));
+                }
+                None => {
+                    self.pending.insert(index, Pending::Unverified(epoch_map));
+                }
+            }
```

Comment:
nit: could possibly be clearer (?)
```suggestion
            let pending = match our_digest {
                Some(digest) => Pending::Verified(digest, epoch_map),
                None => Pending::Unverified(epoch_map),
            };
            self.pending.insert(index, pending);
```

Reply:
Now that we also call `set_rebroadcast_deadline`, it is clearer as is IMO (but would've made sense if not)